src/bastion/src/actor/mailbox.rs

```diff
@@ -2,148 +2,199 @@ use crate::actor::actor_ref::ActorRef;
+            system_rx,
+            latest_user_message,
+            latest_system_message,
+            state,
+        }
     }
 
     /// Forced receive message from user queue
-    pub async fn recv(&self) -> Envelope<T> {
-        self.inner
-            .user_rx()
+    pub async fn recv(&mut self) -> Envelope<T> {
+        if let Some(old_message) = &self.latest_user_message {
+            old_message.ack().await;
+        }
+
+        let message = self
+            .user_rx
             .recv()
             .await
             .map_err(|e| BError::ChanRecv(e.to_string()))
-            .unwrap()
+            .unwrap();
+
+        self.latest_user_message = Some(message);
+        self.latest_user_message.clone().unwrap()
     }
 
     /// Try receiving message from user queue
-    pub async fn try_recv(&self) -> Result<Envelope<T>> {
-        self.inner
-            .user_rx()
-            .try_recv()
-            .map_err(|e| BError::ChanRecv(e.to_string()))
+    pub async fn try_recv(&mut self) -> Result<Envelope<T>> {
+        if self.latest_user_message.is_some() {
+            return Err(BError::UnackedMessage);
+        }
+
+        match self.user_rx.try_recv() {
+            Ok(message) => {
+                self.latest_user_message = Some(message);
+                Ok(self.latest_user_message.clone().unwrap())
+            }
+            Err(e) => Err(BError::ChanRecv(e.to_string())),
+        }
     }
 
-    /// Forced receive message from user queue
-    pub async fn sys_recv(&self) -> Envelope<T> {
-        self.inner
-            .sys_rx()
+    /// Forced receive message from system queue
+    pub async fn sys_recv(&mut self) -> Envelope<T> {
+        if let Some(old_message) = &self.latest_system_message {
+            old_message.ack().await;
+        }
+
+        let message = self
+            .system_rx
             .recv()
             .await
             .map_err(|e| BError::ChanRecv(e.to_string()))
-            .unwrap()
+            .unwrap();
+
+        self.latest_system_message = Some(message);
+        self.latest_system_message.clone().unwrap()
     }
 
-    /// Try receiving message from user queue
-    pub async fn try_sys_recv(&self) -> Result<Envelope<T>> {
-        self.inner
-            .sys_rx()
-            .try_recv()
-            .map_err(|e| BError::ChanRecv(e.to_string()))
+    /// Try receiving message from system queue
+    pub async fn try_sys_recv(&mut self) -> Result<Envelope<T>> {
+        if self.latest_system_message.is_some() {
+            return Err(BError::UnackedMessage);
+        }
+
+        match self.system_rx.try_recv() {
+            Ok(message) => {
+                self.latest_system_message = Some(message);
+                Ok(self.latest_system_message.clone().unwrap())
+            }
+            Err(e) => Err(BError::ChanRecv(e.to_string())),
+        }
+    }
+
+    /// Returns latest system message
+    pub async fn get_latest_user_message(&self) -> Option<Envelope<T>> {
```

Comment:
`get_last_user_message` maybe ?
would pub(crate) be enough?

Reply:
Fixed